### PR TITLE
perf: compute goodness-of-fit K-S statistic directly (9.2x faster SPI)

### LINKED ISSUE #715: [RFC] Interactive Climate Index Explorer using xarray, Dask, and HoloViz
## Summary

`climate_indices` provides reference implementations of drought and climate indices including SPI, SPEI, PET, PNP, PCI, EDDI, and the Palmer indices.

The library's newer xarray API provides an opportunity to expose these capabilities through an interactive visualization application: users could explore gridded climate datasets on a map, select a location, and compute and visualize climate indices for that location without first writing a custom Python workflow.

This RFC proposes building a lightweight **Interactive Climate Index Explorer** using the existing `climate_indices` xarray interface together with components from the Pangeo/HoloViz ecosystem.

The initial goal is deliberately modest: prove that the library can support an interactive workflow over a realistic gridded climate dataset while keeping the scientific computation cleanly separated from the application layer.

## Motivation

Today, using `climate_indices` with gridded datasets typically involves writing Python code or using the command-line interface to process NetCDF inputs, followed by separate tooling for visualization.

A small interactive application could make the project useful to a broader audience while also serving as a practical demonstration of the xarray API.

A representative workflow would be:

1. Open a gridded climate dataset.
2. Select a variable and time period.
3. Explore the data spatially on an interactive map.
4. Click a grid cell or geographic location.
5. Extract the corresponding time series with xarray.
6. Compute an index such as SPI or SPEI using `climate_indices`.
7. Display the resulting index time series interactively.

This would provide value both as a user-facing tool and as a real-world integration test for the xarray/Dask API.

## Design Principles

The implementation should follow several principles:

* **Use public `climate_indices` APIs.** The dashboard should behave like an external consumer of the package rather than depend on internal implementation details.
* **Keep computation independent of the UI.** Dataset selection, point extraction, parameter validation, and index computation should be testable without starting a web application.
* **Keep dashboard dependencies optional.** Visualization and web-framework packages should not become required dependencies of the core `climate_indices` package.
* **Support local data first, cloud data naturally.** NetCDF and Zarr should use the same xarray-based data access model wherever practical.
* **Prefer lazy computation where it provides a measurable benefit.** Dask chunking should be chosen according to dataset storage layout and access pattern rather than prescribing a single chunking scheme.
* **Start small enough to validate interactively.** Distributed computing, authentication, user accounts, persistent jobs, and other platform concerns are outside the initial scope.

## Proposed Initial Architecture

The application would use a Pangeo-inspired scientific Python stack:

### Data access

**xarray** provides the primary abstraction for opening and manipulating gridded datasets.

Initial sources may include:

* local NetCDF files;
* local or remote Zarr stores; and
* selected public cloud-hosted climate datasets.

**Dask** can provide lazy and parallel computation where dataset size or storage layout warrants it.

Chunking strategies should be dataset- and workload-aware. In particular, spatial subsetting should occur before loading or rechunking a long temporal series whenever possible.

### Climate-index computation

The application should call the public `climate_indices` xarray adapter for computations such as:

* SPI;
* SPEI;
* PET; and
* additional supported indices where appropriate.

The dashboard should contain as little scientific logic as possible.

Conceptually:

```text
xarray.DataArray
        │
        ▼
 spatial/temporal selection
        │
        ▼
 climate_indices xarray API
        │
        ▼
 result DataArray
```

This separation should allow the same computation functions to be used from the dashboard, notebooks, scripts, and tests.

### Interactive application

For the initial implementation, **Panel** is the proposed application framework.

The visualization stack would initially use:

* Panel — application layout and reactive controls;
* hvPlot — high-level xarray visualization;
* HoloViews — reactive plot composition and event streams;
* GeoViews where more advanced geographic functionality is required; and
* Bokeh as the interactive plotting backend.

Alternative frameworks can be reconsidered if the prototype identifies significant limitations.

## Proposed Interaction Model

```text
┌─────────────────────────────────────┐
│     NetCDF / Zarr Climate Data      │
└──────────────────┬──────────────────┘
                   │
                   ▼
┌─────────────────────────────────────┐
│          xarray + optional Dask     │
│                                     │
│  variable/time/spatial selection    │
└──────────────────┬──────────────────┘
                   │
          ┌────────┴────────┐
          ▼                 ▼
┌──────────────────┐  ┌─────────────────────┐
│ Interactive Map  │  │ Selected Grid Cell  │
│                  │  │ Time Series         │
└────────┬─────────┘  └──────────┬──────────┘
         │ click                 │
         └───────────────────────┘
                                │
                                ▼
                    ┌──────────────────────┐
                    │   climate_indices    │
                    │   xarray API         │
                    └──────────┬───────────┘
                               │
                               ▼
                    ┌──────────────────────┐
                    │ Index Time-Series    │
                    │ Visualization        │
                    └──────────────────────┘
```

## MVP User Experience

The initial application should provide a simple two-panel interface.

### Map

Display a selected time slice of a gridded climate variable, with:

* geographic coordinates;
* coastline or map context where useful;
* scientifically appropriate color mapping;
* time-step selection; and
* clickable grid cells or coordinates.

### Point selection

When the user clicks the map:

1. determine the nearest grid cell;
2. select it using xarray;
3. display its coordinates and basic metadata; and
4. extract its temporal series.

For rectilinear grids this may initially use logic equivalent to:

```python
dataset.sel(lat=lat, lon=lon, method="nearest")
```

The implementation should not assume that all future datasets use identical coordinate names or grid structures.

### Index calculation

The user should be able to choose an index and relevant parameters such as:

* SPI or SPEI;
* accumulation scale, e.g. 1, 3, 6, or 12 months;
* probability distribution where applicable;
* calibration start year; and
* calibration end year.

Changing parameters should recompute the index for the selected location and update the time-series plot.

### Phase 1 — Vertical Slice

Build the smallest end-to-end application demonstrating:

```text
dataset → map → click → xarray series → climate_indices → plot
```

Tasks:

* create a standalone dashboard application outside the core package;
* select a small but realistic gridded climate dataset;
* render one meteorological variable on an interactive map;
* capture a geographic point selection;
* extract the corresponding xarray time series;
* compute one index, initially SPI; and
* display the resulting time series alongside the map.

### MVP Acceptance Criteria

The first milestone is complete when a user can:

1. start the application locally with one documented command;
2. see a real gridded precipitation dataset on a map;
3. click a location;
4. select an SPI accumulation scale;
5. calculate SPI using the public `climate_indices` xarray API; and
6. see the resulting time series update interactively.

No external notebook or preprocessing script should be required for the demonstrated dataset.

### Phase 2 — Generalize the Application

Once the vertical slice works:

* add SPEI and other suitable indices;
* add variable and dataset selection;
* expose calibration-period and distribution controls;
* handle missing values and insufficient calibration data gracefully;
* improve CF metadata and coordinate handling;
* support local NetCDF and remote Zarr sources through a common data-loading interface; and
* add caching where profiling shows it provides useful latency improvements.

### Phase 3 — Reproducibility, Testing, and Deployment

After the application architecture has stabilized:

* add automated tests for data selection and computation;
* add smoke tests for application startup;
* add targeted UI tests for critical interactions;
* provide a container image or Dockerfile;
* document deployment options;
* demonstrate at least one remotely hosted cloud-optimized dataset; and
* evaluate a public demonstration deployment.

Public hosting should be treated as a deployment concern rather than a requirement of the core library.

## Initial Non-Goals

The first implementation does **not** need to provide:

* a general-purpose GIS application;
* arbitrary remote dataset URLs;
* distributed Dask clusters;
* user authentication;
* persistent user sessions;
* asynchronous job queues;
* pre-computation of global index datasets;
* support for every `climate_indices` algorithm;
* support for every grid topology; or
* a production-scale hosted service.

These capabilities can be evaluated after the core interactive workflow has been demonstrated.

## Repository / Packaging Proposal

For the initial implementation, I propose keeping the application **outside `src/climate_indices/`** so that it remains a consumer of the public package API.

Possible locations include:

```text
dashboard/
```

or:

```text
examples/dashboard/
```

Dashboard-specific dependencies should be installed through an optional dependency group rather than becoming required dependencies of the scientific library, for example:

```text
pip install climate-indices[dashboard]
```

A separate repository should only become necessary if the application eventually develops an independent release lifecycle or substantially different infrastructure requirements.

## Questions for Contributors

Feedback is especially welcome on:

1. **Dataset:** What real-world dataset would make the best first demonstration and benchmark? PRISM, CHIRPS, nClimGrid, ERA5, or another source?
2. **Initial index:** Is SPI the right first end-to-end implementation, followed by SPEI?
3. **Application location:** Should the prototype live under `dashboard/`, `examples/dashboard/`, or another top-level location?
4. **Framework:** Are there concrete requirements that would make Panel/HoloViz unsuitable for the first implementation?
5. **Use cases:** Which interactive climate-index workflows would provide the most value to existing `climate_indices` users?

### PR DESCRIPTION
**Pull request by monocongo**
## Summary

Two related changes: a library performance fix, and the research record that made it worth measuring.

### `perf`: goodness-of-fit K-S statistic computed directly

The gamma and Pearson III goodness-of-fit checks called `scipy.stats.kstest` once per calendar time step — 12 calls
per grid cell for monthly data. Profiling SPI showed **93% of its runtime inside these checks**, nearly all of it
scipy's argument-dispatch machinery rather than the test itself.

This computes the K-S D statistic directly from the sorted sample and the fitted CDF, and evaluates the exact
p-value only for steps whose D exceeds the critical value for the threshold — the uncommon case. The critical value
depends only on sample size, so it is cached.

SPI-6 monthly over 480 timesteps:

| | per cell | 144×144 grid | CONUS nClimGrid (~825k cells) |
|---|---|---|---|
| before | 2.07 ms | 43 s | ~28 min |
| after | **0.22 ms** | **4.6 s** | **~3 min** |

This speeds up the CLI, the pipeline and the test suite, not just gridded work.

### Behavior is unchanged

Reported p-values and poor-fit decisions match `scipy.stats.kstest` exactly. Verified over 800 randomized samples
across both distributions with sample sizes from 8 to 60, mixing well-fitting and deliberately ill-fitting data —
zero mismatches in either the decision or the p-value. Four new tests pin this, including a firm assertion that a
clearly poor fit still emits `GoodnessOfFitWarning` (the existing test for that was a soft one).

### `docs`: Climate Engine and the Explorer decision

The 2026-08-08 landscape review concluded that no turnkey tool offers parameterized `climate_indices`-style
computation joined to map/point exploration. It omitted [Climate Engine](https://www.climateengine.org/), whose
public API computes SPI on demand with a caller-supplied `distribution` (`gamma`, `loglogistic`, `nonparametric`)
and a caller-supplied `start_year`/`end_year` standardization period. The conclusion was wrong.

The doc now carries a Climate Engine section with that evidence, a corrected executive conclusion, and the decision
to shelve the Explorer — including the measurement above, which showed the planned whole-cube materialization
architecture was sized for a bottleneck that turned out to be a diagnostic. Superseded options are retained rather
than deleted so the reasoning stays auditable.

Closes #715, #723, #729, #730, #731, #732, #733, #734, #735.

## Verification

```
uv run pytest          # 1069 passed, 37 deselected
uv run ruff check      # All checks passed
uv run mypy src/climate_indices/compute.py   # Success
```

## Summary by Sourcery

Speed up SPI and related computations by replacing unconditional SciPy K-S checks with an equivalent direct goodness-of-fit evaluation, while documenting the resulting Explorer decision and updated landscape research.

Enhancements:
- Optimize gamma and Pearson III goodness-of-fit checks by computing the K-S statistic directly and only invoking exact p-value calculations for potential poor fits.
- Cache K-S critical values by sample size while preserving existing SciPy-compatible fit decisions and p-values.
- Adjust SPI and SPEI benchmark budgets to account for the substantially faster numerical baselines.

Documentation:
- Add a research review of interactive climate-index exploration tools, correct the Climate Engine assessment, and document the decision to shelve the Explorer effort.

Tests:
- Add parity coverage confirming direct K-S goodness-of-fit results match SciPy across distributions, sample sizes, dtypes, and threshold edge cases.
- Strengthen poor-fit warning tests for gamma and Pearson III distributions.

## Summary by CodeRabbit

- **Documentation**
  - Added a comprehensive comparison of interactive climate-index exploration tools and platform capabilities.
  - Documented supported and unsupported climate-index workflows, integration options, validation questions, and relevant references.
  - Updated conclusions based on existing tools and recorded the decision to shelve the Explorer.

- **Improvements**
  - Improved goodness-of-fit evaluation for Gamma and Pearson Type III distributions.
  - Enhanced consistency with established statistical results, including more accurate handling of borderline cases and aggregated poor-fit warnings.

### CODE REVIEW

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/818
- **Lanes run:** 9 successful — numerical correctness, performance/edge cases, tests/API contracts, documentation/research; two quota-failed rechecks were replaced by successful `scout` and `spec-reviewer` lanes.
- **Validation:** `uv run --python 3.13 ruff check src/ tests/` — pass; `uv run --python 3.13 ruff format --check src/ tests/` — pass; `uv run --python 3.13 pytest` — 1074 passed, 37 deselected; Python 3.10 targeted warning tests — 37 passed; `git diff --check base...HEAD` — pass. `uv run --python 3.13 mypy src/` remains blocked before project analysis by NumPy 2.5.3 stubs using a type statement while project mypy targets Python 3.10; the current worktree also has unrelated uncommitted dependency changes.
- **Fix commits:** 9d7708f6 — fix(compute): preserve K-S goodness-of-fit parity; c4dedb1c — docs(research): correct Climate Engine evidence; b7358ca5 — fix(compute): match K-S thresholds across SciPy versions; 815e5370 — fix(compute): preserve K-S float32 boundary parity; b553cf88 — fix(compute): cover K-S float32 rank rounding

### Findings and fix status
#### [P1] Climate Engine was described as `climate_indices`-backed — Fixed
- **Location:** `docs/research/interactive-climate-explorer-landscape.md` (lines 18–20, 242–244)
- **Origin:** Documentation/research lane
- **Problem:** The hosted, limited Climate Engine API was presented as equivalent to local `climate_indices` computation, supporting an overbroad Explorer-shelving conclusion.
- **Applied fix:** Described the matching hosted gamma-SPI controls without claiming library identity; commit `c4dedb1c`.

#### [P2] Reduced-precision K-S checks could disagree with SciPy — Fixed
- **Location:** `src/climate_indices/compute.py` (lines 935–949)
- **Origin:** Numerical correctness and post-fix validation lanes
- **Problem:** Float32 gamma CDF precision, SciPy-version p-value dtype handling, and float32 rank rounding near 0.05 could suppress or add `GoodnessOfFitWarning`s.
- **Applied fix:** Evaluate Gamma CDF in float64; retain direct clear-good-fit shortcut; defer near-threshold cases to `scipy.stats.kstest` within input machine epsilon. Added float16, float32, critical-value, SciPy 1.15, and SciPy 1.18 parity coverage; commits `9d7708f6`, `b7358ca5`, `815e5370`, `b553cf88`.

#### [P2] Pearson poor-fit warning test was non-asserting — Fixed
- **Location:** `tests/test_data_quality_warnings.py` (lines 372–394)
- **Origin:** Tests/API-contracts lane
- **Problem:** The existing conditional test passed even if Pearson goodness-of-fit warnings stopped emitting.
- **Applied fix:** Replaced it with a deterministic bimodal case requiring one aggregated Pearson warning for all 12 time steps; commit `9d7708f6`.

#### [P3] Climate Engine distribution citation was not a link — Fixed
- **Location:** `docs/research/interactive-climate-explorer-landscape.md` (line 323)
- **Origin:** Documentation/research lane; CodeRabbit Comment 1
- **Problem:** `[CE-API-DIST]` had no destination URL, so rendered citations were not clickable.
- **Applied fix:** Added the validated `/raster/mapid/standard_index` endpoint as the reference destination; commit `c4dedb1c`.

#### [P3] Research prose used noncanonical Timescale terminology — Fixed
- **Location:** `docs/research/interactive-climate-explorer-landscape.md` (lines 18, 36, 123, 270)
- **Origin:** Documentation/research lane
- **Problem:** Documentation used `scale`/`scales` where repository language requires Timescale.
- **Applied fix:** Replaced domain usages with Timescale/Timescales; commit `c4dedb1c`.

#### [P4] Markdown metadata produced trailing-whitespace diff failures — Fixed
- **Location:** `docs/research/interactive-climate-explorer-landscape.md` (lines 3–5)
- **Origin:** Repository-hygiene check
- **Problem:** Markdown hard breaks made `git diff --check` fail.
- **Applied fix:** Used a Markdown list for metadata; commit `c4dedb1c`.

### Remaining blockers
None. Local mypy validation is blocked in the current worktree; no P0/P1 issue remains.

File: docs/research/interactive-climate-explorer-landscape.md [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+- **Google Earth Engine** supplies map-click callbacks, image-collection time-series charts, deferred server execution, and hosted Apps, so the interaction can be assembled there with custom code. [EE-CLICK] [EE-CHART] [EE-DEFERRED] [EE-APPS]
+
+The original review then claimed that no turnkey application exposes the exact `climate_indices` scientific controls—SPI, scale 6, gamma distribution, and explicit calibration years—over a computed gridded result with linked map/time-series exploration.
+
+**That claim was wrong, and the 2026-08-08 review missed the tool that disproves it.** Climate Engine's public API computes SPI on demand with a caller-supplied `distribution` and a caller-supplied `start_year`/`end_year` standardization period, distinct from the display window, and its viewer provides the linked map and point series. [CE-API-DIST] [CE-API-SPEC] The reviewed candidate list omitted Climate Engine entirely, so the "not found" finding reflected the gap in the survey rather than a gap in the field.
```
[the remark this answers is not in the document]
Comment (monocongo) [reply docs/research/interactive-climate-explorer-landscape.md]: Resolved by the narrowed wording now in the doc: the shared-`accumulation=6` claim was removed. The Climate Engine section now states only what the cited evidence shows — caller-supplied `distribution` ([CE-API-DIST]) and caller-supplied `start_year`/`end_year` standardization period with linked map and point-series capabilities ([CE-API-SPEC]) — i.e., the separate parameterized map and point-series capabilities this thread asked for.

File: src/climate_indices/compute.py [hunk 2]
```diff
@@ -898,6 +900,55 @@ def _check_calibration_data_quality(
...
+def _ks_poor_fit_p_value(
+    sorted_values: np.ndarray,
+    cdf_values: np.ndarray,
+) -> float | None:
+    """Kolmogorov-Smirnov p-value for a sample, returned only when the fit is poor.
+
+    The D statistic is computed directly rather than through ``scipy.stats.kstest``,
+    whose argument-dispatch machinery dominates the runtime of this check when it runs
+    once per grid cell. The exact p-value is evaluated only for samples within one
+    input-dtype machine epsilon of the critical D value, which is the uncommon case.
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/compute.py]: Applied in 66a34aab: the docstring now states that clearly acceptable fits skip the exact p-value calculation while candidate poor fits and near-threshold values defer to `scipy.stats.kstest`.
